File: grooves-core/src/main/java/com/github/rahulsom/grooves/logging/IndentedLogging.java
package com.github.rahulsom.grooves.logging;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.slf4j.LoggerFactory;

/**
 * AspectJ aspect that provides indented logging for methods annotated with @Trace.
 * This creates a hierarchical trace output with method calls and returns.
 */
@Aspect
public class IndentedLogging {
    private static final ThreadLocal<Integer> indentLevel = new ThreadLocal<>();
    private static final int INITIAL_INDENT = 1;

    /**
     * Increases the current indentation level for nested method calls.
     */
    public static void stepIn() {
        indentLevel.set(getIndentLevel() + 1);
    }

    /**
     * Decreases the current indentation level when returning from method calls.
     */
    public static void stepOut() {
        indentLevel.set(getIndentLevel() - 1);
    }

    /**
     * Returns the current indentation string based on the current nesting level.
     *
     * @return a string of spaces representing the current indentation level
     */
    public static String indent() {
        return " ".repeat(getIndentLevel() * 2);
    }

    private static int getIndentLevel() {
        var level = indentLevel.get();
        return level != null ? level : INITIAL_INDENT;
    }

    private static String eventsToString(List<?> list) {
        return "<... " + list.size() + " item(s)>";
    }

    /**
     * AspectJ advice that intercepts methods annotated with @Trace to provide indented logging.
     * Logs method entry with parameters and method exit with return values or exceptions.
     *
     * @param joinPoint the intercepted method call
     * @param trace the @Trace annotation instance
     * @return the result of the intercepted method call
     * @throws Throwable any exception thrown by the intercepted method
     */
    @SuppressWarnings({"unused", "UnusedParameters"})
    @Around(value = "@annotation(trace)", argNames = "joinPoint,trace")
    public Object around(ProceedingJoinPoint joinPoint, Trace trace) throws Throwable {
        var signature = joinPoint.getSignature();
        var classWithFunction = joinPoint.getTarget().getClass();
        var loggerName = classWithFunction.getName().replaceAll("\\$\\$Lambda.*$", "");
        var log = LoggerFactory.getLogger(loggerName);

        var methodName = signature.getName().equals("invoke")
                ? signature.getDeclaringType().getSimpleName().substring(0, 1).toLowerCase()
                        + signature.getDeclaringType().getSimpleName().substring(1)
                : signature.getName();

        var args = Arrays.stream(joinPoint.getArgs())
                .map(arg -> arg instanceof List<?> ? eventsToString((List<?>) arg) : arg)
                .map(String::valueOf)
                .collect(Collectors.joining(", "));

        log.trace("{}{}({})", indent(), methodName, args);
        stepIn();

        try {
            var result = joinPoint.proceed();
            stepOut();
            var listRender = result instanceof List<?> ? eventsToString((List<?>) result) : result;
            log.trace("{}{}({}) --> {}", indent(), methodName, args, listRender);
            return result;
        } catch (Throwable t) {
            stepOut();
            log.trace("{}{}({}) ~~> {}", indent(), methodName, args, t.getMessage(), t);
            throw t;
        }
    }
}

File: grooves-core/src/main/kotlin/com/github/rahulsom/grooves/logging/IndentedLogging.kt
package com.github.rahulsom.grooves.logging

import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.slf4j.LoggerFactory
import kotlin.concurrent.getOrSet

@Aspect
class IndentedLogging {
    companion object {
        private val indentLevel = ThreadLocal<Int>()
        private const val INITIAL_INDENT = 1
        fun stepIn() = indentLevel.set(indentLevel.getOrSet { INITIAL_INDENT } + 1)
        fun stepOut() = indentLevel.set(indentLevel.getOrSet { INITIAL_INDENT } - 1)

        @JvmStatic
        fun indent() = "".padStart(indentLevel.getOrSet { INITIAL_INDENT } * 2)
        private fun eventsToString(it: List<*>): Any = "<... ${it.size} item(s)>"
    }

    @Suppress("unused", "UNUSED_PARAMETER")
    @Around(value = "@annotation(trace)", argNames = "trace")
    @ExperimentalStdlibApi
    fun around(joinPoint: ProceedingJoinPoint, trace: Trace): Any? {
        val signature = joinPoint.signature
        val classWithFunction = joinPoint.target.javaClass
        val loggerName = classWithFunction.name.replace(Regex("\\\$\\\$Lambda.*$"), "")
        val log = LoggerFactory.getLogger(loggerName)
        val methodName =
            if (signature.name == "invoke") {
                signature.declaringType.simpleName.replaceFirstChar { x -> x.lowercaseChar() }
            } else {
                signature.name
            }

        val args = joinPoint.args.map { if (it is List<*>) eventsToString(it) else it }.joinToString(", ")
        log.trace("${indent()}$methodName($args)")
        stepIn()

        try {
            val result = joinPoint.proceed()
            stepOut()
            val listRender = if (result is List<*>) eventsToString(result) else result
            log.trace("${indent()}$methodName($args) --> $listRender")
            return result
        } catch (t: Throwable) {
            stepOut()
            log.trace("${indent()}$methodName($args) ~~> $t")
            throw t
        }
    }
}